File: src/murmur/compaction.py
"""Periodic profile compaction, off the live loop (spec 05 §3.6).

The Compactor watches the persistent store's backlog and folds it into the
long-term profile via the Brain seam — as a single-flight background task, so
it never blocks a segment. Failure is inert: the profile and watermark are left
untouched (the store only advances them on a successful apply), one dev-log
line, and the backlog is retried next time. A stub Brain makes the fold a
no-op (its ``compact_profile`` returns the profile unchanged), so an offline
run costs nothing.
"""

from __future__ import annotations

import asyncio
from typing import Protocol

from .contracts import Turn
from .logging_setup import get_log

_log = get_log("memory")


class _CompactStore(Protocol):
    def compaction_due(self) -> bool: ...
    def compaction_slice(self) -> tuple[str, list[Turn], float]: ...
    def apply_compaction(self, new_profile: str, *, through_ts: float) -> None: ...


class _CompactBrain(Protocol):
    async def compact_profile(self, profile: str, transcript: list[Turn]) -> str: ...


class Compactor:
    """Single-flight scheduler around a store's compaction surface."""

    def __init__(self, store: _CompactStore, brain: _CompactBrain) -> None:
        self._store = store
        self._brain = brain
        self._task: asyncio.Task[None] | None = None

    def maybe_schedule(self) -> bool:
        """If the backlog crossed the threshold and nothing is in flight, launch
        one background compaction. Returns whether it launched one."""
        if not self._store.compaction_due():
            return False
        return self._launch()

    async def flush(self) -> None:
        """Force a compaction of any remaining backlog, regardless of the
        threshold (shutdown / startup catch-up). Drains a fold already in flight
        first — its slice predates the current tail — then folds what is left, so
        turns recorded during that in-flight fold are not stranded until a future
        run. Best-effort (the fold swallows its own errors); bounded to two
        rounds (the settled fold + the tail) so a persistently-failing fold can
        never spin here."""
        await self.drain()  # let an in-flight fold (older slice) finish first
        _, turns, _ = self._store.compaction_slice()
        if turns:
            self._launch()
            await self.drain()

    async def drain(self) -> None:
        """Await the in-flight compaction, if any (shutdown / tests)."""
        if self._task is not None:
            await self._task
            self._task = None

    def _launch(self) -> bool:
        if self._task is not None and not self._task.done():
            return False  # single-flight: one fold at a time
        self._task = asyncio.ensure_future(self._run())
        return True

    async def _run(self) -> None:
        profile, turns, through_ts = self._store.compaction_slice()
        if not turns:
            return
        try:
            updated = await self._brain.compact_profile(profile, turns)
        except Exception as exc:  # noqa: BLE001 - never let compaction crash the radio
            _log.warn("compaction failed; keeping profile + watermark", exc=exc)
            return
        self._store.apply_compaction(updated, through_ts=through_ts)
        _log.event("memory.compacted", turns=len(turns), profile_chars=len(updated))

File: src/compaction.ts
// Periodic profile compaction, off the live loop (spec 05 §3.6).
//
// The Compactor watches the persistent store's backlog and folds it into the
// long-term profile via the Brain seam — single-flight and in the background,
// so it never blocks a segment. Failure is inert: the store only advances
// profile + watermark on a successful apply, so a failed fold costs one log
// line and the backlog is retried next time. A stub Brain makes the fold a
// no-op (its compactProfile returns the profile unchanged).

import type { Brain, Turn } from './contracts.ts'

// The store's compaction surface (spec 05 §2.1) — impl-level, deliberately not
// on the MemoryStore contract: the Director never drives compaction.
export interface CompactionStore {
  compactionDue(): boolean
  compactionSlice(): { profile: string; turns: Turn[]; throughTs: number }
  applyCompaction(newProfile: string, throughTs: number): void
}

type Fold = { promise: Promise<void>; done: () => boolean }

// How long a shutdown flush may wait on the model before it gives up. A fold
// is a full model call — measured at 53 s on a real run — so this is not a
// generous version of "wait for it": it is only long enough to collect a fold
// that was already nearly done, and Ctrl-C must feel like stopping.
const FLUSH_TIMEOUT_MS = 3_000

export class Compactor {
  private store: CompactionStore
  private brain: Pick<Brain, 'compactProfile'>
  private log: (message: string) => void
  private fold: Fold | null = null

  constructor(
    store: CompactionStore,
    brain: Pick<Brain, 'compactProfile'>,
    log: (message: string) => void = () => {},
  ) {
    this.store = store
    this.brain = brain
    this.log = log
  }

  // If the backlog crossed the threshold and nothing is in flight, launch one
  // background fold. Returns whether it launched one.
  maybeSchedule(): boolean {
    if (!this.store.compactionDue()) return false
    return this.launch()
  }

  // Force a fold of any remaining backlog, regardless of the threshold
  // (shutdown / startup catch-up). Drains a fold already in flight first — its
  // slice predates the current tail — then folds what is left, so turns
  // recorded during that fold are not stranded until a future run.
  // Bounded two ways, so a persistently-failing fold can never spin here and a
  // slow one can never hold the process open: two rounds, and a wall-clock
  // budget. Losing the fold costs nothing — the turns are already on disk and
  // the watermark only advances on a successful apply, so the next run folds
  // them again. Overrunning the budget abandons the wait, not the fold: a
  // promise cannot be cancelled, so it keeps running — and its SDK subprocess
  // does not hold the process open, because main.ts exits explicitly rather
  // than waiting for the loop to drain (the same disposal the in-flight music
  // pick already relies on; measured: 3.5 s from Ctrl-C to exit).
  async flush(timeoutMs = FLUSH_TIMEOUT_MS): Promise<void> {
    await Promise.race([
      this.foldRemaining(),
      // .unref(): the bound must never be the thing keeping the loop alive.
      new Promise<void>((resolve) => setTimeout(resolve, timeoutMs).unref()),
    ])
  }

  private async foldRemaining(): Promise<void> {
    await this.drain()
    if (this.store.compactionSlice().turns.length > 0) {
      this.launch()
      await this.drain()
    }
  }

  // Await the in-flight fold, if any (shutdown / tests).
  async drain(): Promise<void> {
    await this.fold?.promise
    this.fold = null
  }

  private launch(): boolean {
    if (this.fold !== null && !this.fold.done()) return false
    let settled = false
    this.fold = {
      promise: this.run().finally(() => (settled = true)),
      done: () => settled,
    }
    return true
  }

  // Total (never rejects): the fold runs unawaited in the background, so any
  // escape here — Brain failure OR a filesystem error in apply — would be an
  // unhandled rejection that takes the radio down instead of degrading.
  private async run(): Promise<void> {
    const { profile, turns, throughTs } = this.store.compactionSlice()
    if (turns.length === 0) return
    try {
      const updated = await this.brain.compactProfile(profile, turns)
      this.store.applyCompaction(updated, throughTs)
      this.log(`memory: compacted ${turns.length} turns into a ${updated.length}-char profile`)
    } catch (err) {
      this.log(`memory: compaction failed; keeping profile + watermark (${String(err)})`)
    }
  }
}
